src/OptimizedPathMappingRepository.php

```diff
@@ -0,0 +1,464 @@
+    {
+        $staticPrefix = rtrim($path, '/').'/';
+        $regExp = '~^'.preg_quote($staticPrefix, '~').'.+$~';
+
+        return new RegexFilterIterator(
+            $regExp,
+            $staticPrefix,
+            new ArrayIterator($this->store->keys())
+        );
+    }
+
+    /**
+     * Returns an iterator for a glob.
+     *
+     * @param string $glob The glob.
+     *
+     * @return GlobFilterIterator|string[] The iterator of paths.
+     */
+    private function getGlobIterator($glob)
+    {
+        return new GlobFilterIterator(
+            $glob,
+            new ArrayIterator($this->store->keys())
+        );
+    }
+
+    /**
+     * Recursively creates a directory for a path.
+     *
+     * @param string $path A directory path.
+     * @return DirectoryResource The created resource
+     */
+    private function ensureDirectoryExists($path)
+    {
+        if (!$this->store->exists($path)) {
+            // Recursively initialize parent directories
+            if ($path !== '/') {
+                $this->ensureDirectoryExists(Path::getDirectory($path));
+            }
+
+            $this->store->set($path, null);
+
+            return;
+        }
+    }
+
+    /**
+     * Transform an iterator of paths into a collection of resources
+     *
+     * @param \Iterator $iterator
+     * @return FilesystemResourceCollection
+     */
+    private function iteratorToCollection(\Iterator $iterator)
+    {
+        $filesystemPaths = $this->store->getMultiple(iterator_to_array($iterator));
+        $resources = array();
+
+        foreach ($filesystemPaths as $path => $filesystemPath) {
+            $resource = $this->unserialize($filesystemPath);
+            $resource->attachTo($this, $path);
+
+            $resources[] = $resource;
+        }
+
+        return new FilesystemResourceCollection($resources);
+    }
+
+    /**
+     * Count the number of elements in the store
+     */
+    private function createRoot()
+    {
+        if ($this->store->exists('/')) {
+            return;
+        }
+
+        $this->store->set('/', null);
+    }
+
+    /**
+     * Count the number of elements in the store
+     */
+    private function countStore()
+    {
+        if ($this->store instanceof Countable) {
+            return count($this->store);
+        }
+
+        return count($this->store->keys());
+    }
+
+    /**
+     * Sort the store by keys
+     */
+    private function sortStore()
+    {
+        $resources = $this->store->getMultiple($this->store->keys());
+
+        ksort($resources);
+
+        $this->store->clear();
+
+        foreach ($resources as $path => $resource) {
+            $this->store->set($path, $resource);
+        }
+    }
+
+    /**
+     * Create a string representing a given resource for the store
+     *
+     * @param FilesystemResource $resource
+     * @return string
+     */
+    private function serialize(FilesystemResource $resource)
```

Comment:
I don't think this method has any benefit. I'd use $resource->getFilesystemPath() right away, then it's obvious what is happening. At the moment, you could believe that the method serializes the resource.